sklearn/cluster/optics_.py

```diff
@@ -527,3 +555,379 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+        constitutes a cluster boundary. For example, an upwards point in the
+        reachability plot is defined by the ratio from one point to its
+        successor being at most 1-xi.
+
+    predecessor_correction : bool, optional (default=True)
+        Correct clusters based on the calculated predecessors.
+
+    Returns
+    -------
+    labels : array, shape (n_samples)
+        The labels assigned to samples. Points which are not included
+        in any cluster are labeled as -1.
+
+    clusters : array, shape (n_clusters, 2)
+        The list of clusters in the form of ``[start, end]`` in each row, with
+        all indices inclusive. The clusters are ordered according to
+        ``(end, -start)`` (ascending) so that larger clusters encompassing
+        smaller clusters come after those smaller ones. Since ``labels`` does
+        not reflect the hierarchy, usually
+        ``len(clusters) > np.unique(labels)``.
+    """
+    clusters = _xi_cluster(reachability[ordering], predecessor[ordering], xi,
+                           min_samples, min_cluster_size,
+                           predecessor_correction)
+    labels = _extract_xi_labels(ordering, clusters)
+    return labels, clusters
+
+
+class _Area:
+    """An (upward or downward) area.
+
+    Attributes
+    ----------
+    start : integer
+        The start of the region.
+
+    end : integer
+        The end of the region.
+
+    maximum : float
+        The maximum reachability in this region, which is the
+        start of the region for a downward area and the end of
+        the region for an upward area.
+
+    mib : float
+        Maximum in between value, i.e. the maximum value between
+        the end of a steep down area and the current index. It is
+        irrelevant for steep up areas.
+    """
+    def __init__(self, start, end, maximum, mib):
+        self.start = start
+        self.end = end
+        # maximum is the beginning of a downward area
+        # and the end of an upward area
+        self.maximum = maximum
+        # maximum in between (ref Sec. 4.3.2)
+        self.mib = mib
+
+    def __repr__(self):
+        return ("start: %s, end: %s, max: %4g, mib: %4g" %
+                (self.start, self.end, self.maximum, self.mib))
+
+
+def _extend_region(steep_point, xward_point, start, min_samples):
+    """Extend the area until it's maximal.
+
+    It's the same function for both upward and downward reagions, depending on
+    the given input parameters. To extend an upward reagion,
+    ``steep_point=steep_upward`` and ``xward_point=downward`` are expected, and
+    to extend a downward region, ``steep_point=steep_downward`` and
+    ``xward_point=upward``.
+
+    Parameters
+    ----------
+    steep_point : bool array, shape (n_samples)
+        True if the point is an upward or downward steep point, respectively.
+
+    xward_point : bool array, shape (n_samples)
+        True if the point is upward or downward respectively.
```

Comment:
it's not respective, it's oppositely. Maybe write "True if the point is downward (upward)." here and "True if the point is an upward (respectively downward) steep point."

Reply:
In my mind it was respective to whether it was called as extend_upward or downward, not the other parameter. But I'll change it.